### ISSUE TITLE
fix-building-without-safebrowsing.patch incomplete?

### ISSUE THREAD

**Issue by teeminus** [Eloston/ungoogled-chromium#1935]
Hi all,

while updating the windows build I encountered a linking error because the symbol `safe_browsing::EventResultToString` is not found. The function is used multiple times in the file `chrome/browser/extensions/api/safe_browsing_private/safe_browsing_private_event_router.cc`, but most calles are within `FULL_SAFE_BROWSING` ifdef blocks inserted by the `fix-building-without-safebrowsing.patch` patch except for one call around line 313 in function `SafeBrowsingPrivateEventRouter::OnPolicySpecifiedPasswordReuseDetected`.

The missing function is implemented in the file `chrome/browser/safe_browsing/cloud_content_scanning/deep_scanning_utils.cc` which is not compiled because the patch `0001-fix-building-without-safebrowsing.patch` disables compiling of the safe_browsing files defined in `chrome/browser/safe_browsing/BUILD.gn`.

I was wondering if windows is the only platform affected by this problem or if this should be fixed in this repo's patch.

**Comment by squid-f** [comment]
Hi. 
I am facing the same issue on Linux. 
I have tried to change link flags but it didn't help.
The error I get is:
```
[49024/49024] LINK ./chrome
FAILED: chrome 
python3 "../../build/toolchain/gcc_link_wrapper.py" --output="./chrome" -- clang++ -Wl,--version-script=../../build/linux/chrome.map -fuse-ld=lld -Wl,--build-id -fPIC -Wl,-z,noexecstack -Wl,-z,relro -Wl,-z,now -Wl,--icf=all -Wl,--color-diagnostics -Wl,--no-call-graph-profile-sort -no-canonical-prefixes -rdynamic -Wl,-z,defs -Wl,--as-needed -pie -Wl,--disable-new-dtags  -Wl,--as-needed -Wl,--no-undefined -Wl,-z,relro -Wl,-O1 -Wl,--build-id -Wl,--enable-new-dtags -o "./chrome" -Wl,--start-group @"./chrome.rsp"  -Wl,--end-group  -latomic -ldl -lpthread -lrt -lgmodule-2.0 -lglib-2.0 -lgobject-2.0 -lgthread-2.0 -lsmime3 -lnss3 -lsoftokn3 -lnssutil3 -lplds4 -lplc4 -lnspr4 -latk-1.0 -latk-bridge-2.0 -lcups -lgio-2.0 -ldrm -lfontconfig -ldbus-1 -ljsoncpp -lresolv -lz -lpng16 -lwebpdemux -lwebpmux -lwebp -lfreetype -ljpeg -lexpat -lm -lopus -lavcodec -lavformat -lavutil -lX11 -lXcomposite -lXdamage -lXext -lXfixes -lXrender -lXrandr -lXtst -lxcb -lX11-xcb -lxkbcommon -lgbm -lwayland-client -lpangocairo-1.0 -lpango-1.0 -lharfbuzz -lcairo -ludev -lre2 -lXi -lpci -lasound -lsnappy -latspi -lxml2 -lspeechd -lFLAC -lminizip -lxshmfence -lxslt -lopenjp2 -lpng -ltiff
ld.lld: error: undefined symbol: safe_browsing::EventResultToString[abi:cxx11](safe_browsing::EventResult)
>>> referenced by safe_browsing_private_event_router.cc
>>>               extensions/safe_browsing_private_event_router.o:(extensions::SafeBrowsingPrivateEventRouter::OnPolicySpecifiedPasswordReuseDetected(GURL const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool, bool)) in archive obj/chrome/browser/extensions/libextensions.a
```

The upstream version of Chromium does build with the same build options.

I just saw @PF4Public mentioned #139. I will give it a try.
Is there anything else I can test?

**Comment by PF4Public** [comment]
@squid-f What version of clang do you use?

**Comment by squid-f** [comment]
> @squid-f What version of clang do you use?

11.0.1

I gave a try to @teeminus patch, updated for Chromium 101, and the link step now does pass.

**Comment by PF4Public** [comment]
@teeminus , @squid-f by any chance could you try building with clang-14 and clang-15?

**Comment by teeminus** [comment]
I already updated clang to 14.0.2 because otherwise compiling failed due to some missing c++20 features.

The most recent clang version I could use is 14.0.3 (see https://github.com/llvm/llvm-project/releases) but I doubt this would change anything.

**Comment by PF4Public** [comment]
@Ahrotahn @Zoraver Please voice your opinion.

**Comment by jstkdng** [comment]
It compiles without a problem on arch with clang-13.
Could this be a windows problem maybe?

**Comment by PF4Public** [comment]
How about `is_official_build=`? I'm building with `yes`. @teeminus @squid-f @jstkdng you?

**Comment by teeminus** [comment]
I guess for windows the flag should be in the 'flags.windows.gn' file, correct? Currently there is no such flag set there.

**Comment by jstkdng** [comment]
@PF4Public 
This are all the stuff related to the flags on arch
```bash
  local _flags=(
    'custom_toolchain="//build/toolchain/linux/unbundle:default"'
    'host_toolchain="//build/toolchain/linux/unbundle:default"'
    'is_official_build=true' # implies is_cfi=true on x86_64
    'symbol_level=0' # sufficient for backtraces on x86(_64)
    'chrome_pgo_phase=0' # needs newer clang to read the bundled PGO profile
    'disable_fieldtrial_testing_config=true'
    'blink_enable_generated_code_formatting=false'
    'ffmpeg_branding="Chrome"'
    'proprietary_codecs=true'
    'rtc_use_pipewire=true'
    'link_pulseaudio=true'
    'use_gnome_keyring=false'
    'use_sysroot=false'
    'use_custom_libcxx=false'
    'enable_widevine=true'
  )

  if [[ -n ${_system_libs[icu]+set} ]]; then
    _flags+=('icu_use_data_file=false')
  fi

  # Append ungoogled chromium flags to _flags array
  _ungoogled_repo="$srcdir/$pkgname-$_uc_ver"
  readarray -t -O ${#_flags[@]} _flags < "${_ungoogled_repo}/flags.gn"
```

**Comment by squid-f** [comment]
> @teeminus , @squid-f by any chance could you try building with clang-14 and clang-15?

Hi. I just launched a build with clang 14. Let us see.

> How about is_official_build=? I'm building with yes. @teeminus @squid-f @jstkdng you?

I use the following flags:
```
--args=use_sysroot=false 
                  system_libdir="lib64" 
                  icu_use_data_file=false 
                  is_nacl_glibc=false 
                  custom_toolchain="//build/toolchain/linux/unbundle:default" 
                  host_toolchain="//build/toolchain/linux/unbundle:default" 
                  proprietary_codecs=true 
                  linux_link_libudev = true 
                  linux_link_libspeechd = true 
                  linux_link_xlib = true 
                  enable_platform_ac3_eac3_audio = true 
                  enable_platform_dts_audio = true 
                  enable_platform_hevc=true 
                  use_system_libjpeg=true 
                  use_system_libopenjpeg2=true 
                  use_libjpeg_turbo=false 
                  enable_widevine=true 
                  pdf_enable_xfa=true 
                  gtk_version=3 
                  fatal_linker_warnings=false 
                  use_libpci = true 
                  use_pulseaudio=true 
                  use_allocator="none" 
                  is_debug=false 
                  is_clang=true 
                  clang_use_chrome_plugins=false 
                  use_lld=true 
                  use_gold=false 
                  use_system_freetype=true 
                  use_system_harfbuzz=true 
                  use_system_libpng=true 
                  enable_dav1d_decoder=true enable_libaom=false 
                  use_custom_libcxx=false 
                  enable_mojom_closure_compile=false 
                  dcheck_always_on=false 
                  treat_warnings_as_errors=false 
                use_vaapi=true 
                use_system_libwayland=true 
                use_system_wayland_scanner=true 
                is_component_build=false 
                  build_with_tflite_lib=false 
                  chrome_pgo_phase=0 
                  clang_use_chrome_plugins=false 
                  disable_fieldtrial_testing_config=true 
                  enable_hangout_services_extension=false 
                  enable_js_type_check=false 
                  enable_mdns=false 
                  enable_mse_mpeg2ts_stream_parser=true 
                  enable_nacl=false 
                  enable_one_click_signin=false 
                  enable_reading_list=false 
                  enable_remoting=false 
                  enable_reporting=false 
                  enable_service_discovery=false 
                  exclude_unwind_tables=true 
                  google_api_key="" 
                  google_default_client_id="" 
                  google_default_client_secret="" 
                  safe_browsing_mode=0 
                  use_official_google_api_keys=false 
                  use_unofficial_version_number=false
```
is_official_build is not listed. What is the default value? I think I have read somewhere we should not use it as it implies many google stuffs we don't want ?

**Comment by squid-f** [comment]
Hi.
So, using clang 14 instead clang 11 allows passing the link step, without using @teeminus patch; using the same flags as the one listed on my previous post.

**Comment by h0cheung** [comment]
I'm trying to pack this for opensuse based on the chromium package in openSUSE:Factory. The compiler is gcc11 and always got this error.

**Comment by Ahrotahn** [comment]
It looks like it's due to some difference between clang versions.  I'll try to remember to include that removal in one of the next updates I work on.  

The `is_official_build` flag enables link time optimizations, allows for PGO, and also disables `is_debug`.  I recommend building with `is_official_build` when possible since we don't normally check if there are any removals needed inside debug definitions.

**Comment by teeminus** [comment]
I'm currently recompiling with 'is_official_build=true' and my patch removed.

**Comment by h0cheung** [comment]
> It looks like it's due to some difference between clang versions. I'll try to remember to include that removal in one of the next updates I work on.
> 
> The `is_official_build` flag enables link time optimizations, allows for PGO, and also disables `is_debug`. I recommend building with `is_official_build` when possible since we don't normally check if there are any removals needed inside debug definitions.

I tried `is_official_buid`, which seems not related to this.

**Comment by Michal-Szczepaniak** [comment]
Same here while building on opensuse based on chromium build, with patches applied

https://build.opensuse.org/package/live_build_log/home:Mister_Magister/chromium/standard/x86_64

**Comment by teeminus** [comment]
It's fixed with 102.0.5005.61.

**Comment by Michal-Szczepaniak** [comment]
Confirmed, working fine on 102
